### DISCUSSION TITLE
Transform citekeys to find correct notes and library files

### DISCUSSION THREAD

**Discussion by tillheilmann** [emacs-citar/citar#844] [Q&A] [answered]
My BibTeX keys include a colon to join author name and publication year, like this:

`@seideman.j:1999` 

In associated note and library files, the colon is substituted by a period, like this:

`~/notes/seideman.j.1999.txt` and `~/pdf/seideman.j.1999.pdf`

What is the right place in citar functions and variables to make the according changes so the package finds the right .txt and .pdf files?

**Comment by bdarcus** [comment]
> What is the right place in citar functions and variables to make the according changes so the package finds the right .txt and .pdf files?

Been awhile since I've worked with this code, so can't point you to anything specific. But you'd want to look in `citar-file.el`.

Am not sure if we allow that sort of configurability, though; we've always assumed key and file name match. You'd likely have to somehow transform the key before running the functions to get those files ([citar-file--get-library-files](https://github.com/emacs-citar/citar/blob/0f1786b7fee58452a3225e4b9b7c94176fff9b5a/citar-file.el#L223) might be worth looking at).

But please report back what you figure out.

**Comment by tillheilmann** [comment]
Thanks for the hint!

I tried to redefine how the `KEYS` argument is given to `citar-file--directory-files` in `citar-file--get-library-files` (see below), but it didn't work. When calling it with key `X` on point, package tells me there are `No associated files for (X)`.

```
(defun citar-file--get-library-files (&optional keys)
  "Return list of files for KEYS in ENTRIES."
  (citar--check-configuration 'citar-library-paths 'citar-library-file-extensions)
  (citar-file--directory-files
   (citar-file--library-dirs)
   ;; changed ...
   ;; keys
   ;; ... to following for replacing colons with periods in all elements of list.
   (dolist (item keys)
     (string-replace ":" "." item))
   citar-library-file-extensions
   citar-file-additional-files-separator))
```

**Reply by bdarcus** [reply]
Doesn't that code replace all periods, but you only need the last one?

**Reply by tillheilmann** [reply]
According to the docstring of `string-replace`, yes. But that should not matter in my case because there is only ever one colon (to be replaced by a period) in a citekey. If

```
   (dolist (item keys)
     (string-replace ":" "." item))
```

looks correct to you (instead of a simple `keys` in above code), then there maybe something else wrong with my setup.

**Comment by roshanshariff** [comment, marked as answer]
I'm typing from my phone so code is hard, but string-replace returns a new
string instead of modifying the string you pass it. Your code is looping
over the list of strings, modifying each of them, but not doing anything
with the result.

You might want to use a mapcar instead of dolist.

**Reply by tillheilmann** [reply]
Thank you very much! My understanding of Elisp is not even basic, so I effectively do trial-and-errors on copy-pasted code snippets I find on the Internet. However, this seems to work (see below). Does it look clean enough for everyday use?

```
(defun citar-file--get-library-files (&optional keys)
  "Return list of files for KEYS in ENTRIES."
  (citar--check-configuration 'citar-library-paths 'citar-library-file-extensions)
  (citar-file--directory-files
   (citar-file--library-dirs)
   (mapcar (lambda (key) (string-replace ":" "." key)) keys) ; colons to periods in citekeys 
   citar-library-file-extensions
   citar-file-additional-files-separator))
```

**Reply by tillheilmann** [reply]
Sorry, that was premature. What I tried above doesn't work. I still get the same `No associated files for (X)`. (I thought it did work because the citekey I was using to test the modified function was somehow changed to the 'correct' format without a colon.)

**Reply by tillheilmann** [reply]
I think I understand what the problem is: I would probably have to modify `citar-file--directory-files`, not `citar-file--get-library-files`, because this is the function that actually constructs filenames from citekeys.

**Reply by tillheilmann** [reply]
I changed line 302 in `citar-file.el` to

```
(let ((filename (expand-file-name (concat (string-replace ":" "." key) "." ext) dir)))
```

Not pretty. But it seems to do the trick.
